src/primitive_crypto.cc

```diff
@@ -118,6 +120,206 @@ PRIMITIVE(siphash_get) {
+
+  ByteArray* proxy = process->object_heap()->allocate_proxy();
+  if (proxy == null) ALLOCATION_FAILED;
+
+  mbedtls_cipher_id_t mbedtls_cipher;
+
+  if (algorithm == ALGORITHM_AES_GCM_SHA256) {
+    mbedtls_cipher = MBEDTLS_CIPHER_ID_AES;
+  } else {
+    INVALID_ARGUMENT;
+  }
+
+  GcmContext* gcm_context = _new GcmContext(
+      group,
+      mbedtls_cipher,
+      encrypt);
+
+  if (!gcm_context) {
+    MALLOC_FAILED;
+  }
+
+  int err = mbedtls_gcm_setkey(gcm_context->gcm_context(), mbedtls_cipher, key.address(), key.length() * BYTE_BIT_SIZE);
+  if (err != 0) {
+    group->unregister_resource(gcm_context);
+    return tls_error(null, process, err);
+  }
+  
+  proxy->set_external_address(gcm_context);
+  return proxy;
+}
+
+PRIMITIVE(gcm_close) {
+  ARGS(GcmContext, context);
+  context->resource_group()->unregister_resource(context);
+  context_proxy->clear_external_address();
+  return process->program()->null_object();
+}
+
+// Start the encryption of a message, or TLS record.  Takes a 12 byte nonce.
+// As described in RFC5288 section 3, the first 4 bytes of the nonce are kept
+// secret, and the next 8 bytes are transmitted along with each record.
+// Internally these primitives will add 4 more bytes of block counter, starting
+// at 1, to form a 16 byte IV.  It is vital that the nonce is not reused, and
+// this is normally achieved by counting up the explicit part by one for each
+// record that is encrypted.  In TLS this means that this part of the nonce
+// corresponds to the sequence number of the record.
+PRIMITIVE(gcm_start_message) {
+  ARGS(GcmContext, context, int, length, Blob, authenticated_data, Blob, nonce);
+  if (context->remaining_length_in_current_message() != 0) INVALID_ARGUMENT;
+  context->set_remaining_length_in_current_message(length);
+  if (nonce.length() != GcmContext::NONCE_SIZE) INVALID_ARGUMENT;
+  int mode = context->is_encrypt() ? MBEDTLS_GCM_ENCRYPT : MBEDTLS_GCM_DECRYPT;
+  int result = mbedtls_gcm_starts(
+      context->gcm_context(),
+      mode,
+      nonce.address(),
+      nonce.length(),
+      authenticated_data.address(),  // No additional data.
+      authenticated_data.length());
+  if (result != 0) return tls_error(null, process, result);
+
+  return process->program()->null_object();
+}
+
+/**
+If the result byte array was big enough, returns a Smi to indicate how much
+  data was placed in it.
+If the result byte array was not big enough, returns null.  In this case no
+  data was consumed.
+*/
+PRIMITIVE(gcm_add) {
+  ARGS(GcmContext, context, Blob, data, MutableBlob, result);
+  const uint8* data_address = data.address();
+  int data_length = data.length();
+  if (data_length == 0) return Smi::from(0);
+  int remains = context->remaining_length_in_current_message();
+  if (remains < data_length) OUT_OF_BOUNDS;
+  remains -= data_length;
+
+  // Start by copying into the temporary buffer in the context.
```

Comment:
I think I would structure this differently:
1. the old and new data are smaller than a block.
  -> add the new data to the buffered. and return
2. compute the output size (a multiple of the block size). At this stage I would ignore whether this is the last block or not.  If it is the last block, then we might buffer something, which we should then handle in the `finish`/`verify` function.
3. if the result buffer isn't big enough -> return null
4. if buffer is not empty, fill it up and encrypt it.  (There is a good chance the chunks that come in here are aligned by 16 bytes)
5. encrypt the rest that can be handled this time.


Change `finish`/`verify` to crypt the last buffered chunk (if there is any).

Reply:
Not knowing the size up front means that PRIMITIVE(gcm_add) has been rewritten.  I think it's clearer now.

The encrypt method on an instance of the class follows the other Aes classes. :-/